### ISSUE TITLE
Question regarding best practice for mapping optional nested records via row().mapping()

### ISSUE THREAD

**Issue by matt-snider** [jOOQ/jOOQ#12512]
### Your question:

Hi! First of all I'd like to thank everyone who has taken part in creating JOOQ. JOOQ is easily my favourite database layer library across the many I've encountered in multiple programming languages.Many thanks for the incredible work and for providing so much value to the community.

My question centers around optional nested records (e.g. via a left join) and what the best practice would be for mapping them.

This is an example of where this would be used:

```java
row(
    EMPLOYEES.ID.convertFrom(UInteger::longValue),
    EMPLOYEE.NAME,
    EMPLOYEE.TITLE
).mapping(Employee::new)
```

And my current solution is to do this:

```java
row(
    EMPLOYEES.ID.convertFrom(x -> x != null ? x.longValue() : null),
    EMPLOYEE.NAME,
    EMPLOYEE.TITLE
).mapping((t1, t2, t3) -> t1 != null ? new Employee(t1, t2, t3) : null);
```

Which I've since extracted to a utility function, but this still needs to be applied to the `mapping()` call and any field-specific conversions with `convertFrom()` because otherwise we get a NPE.

This solution works, but I'm just wondering if there is a better way. I tried a few things that either didn't compile or failed at runtime, mostly with this sort of approach `if_(EMPLOYEE.ID.isNull(), value(null, Long.class).convertFrom((x) -> (Employee) null), row(...).mapping(...))`, but it ends up being a nested field `Field<SelectField<Employee>>`.

If there isn't currently a better way, I'm wondering if something like `optionalMapping(Employee::new)` or `optionalRow(...)` would work, which would also need to short circuit the field level conversions. 

(By the way, here I also have to give a bit of praise because the new nested record support via `row().mapping()` is awesome! :sunglasses:)

### Versions:

- jOOQ: 3.15.3
- Java: 16
- Database (include vendor): MySQL 8

**Comment by lukaseder** [comment]
https://github.com/jOOQ/jOOQ/issues/12513 is implemented for 3.16.0. You can copy the class to your workspace for the time being: https://github.com/jOOQ/jOOQ/commit/18f9d9dd12cae0f8d23f251e150809c786fd7092

**Comment by lukaseder** [comment]
> so these conversions need to wrapped as well (`FIELD.convertFrom(nullOnAnyNull(converter))`).

Well, I contemplated offering utilities to handle nulls out of the box, after all, there's `Converter.ofNullable()`. The choices were:

- Status quo (least of evils)
- Offer 5 (!) additional methods on `Field`, and also on `DataType`, etc. just for that (to an already huge API)
- Make the `Converter.ofNullable()` behaviour the default (not nice for those edge cases where `null` needs to be wrapped in something non-null, on either the `T` or `U` side of a converter, i.e. it would be hard to opt out)

Other languages don't require clumsy conditional expressions like `a == null ? null : b`, but have special operators for these cases...

> I understand though that if you added new methods for every little idea like that, the API size would grow and become less coherent

This is what happened with all those `fetchMap`, `fetchGroups`, `fetchSet` and other overloads for `Field`, `Name`, `String`, `int`, `RecordMapper`, `Class`, etc. arguments! The solution is more composition, less overloading, in the long run. In that case, via the `Collector` API, see https://github.com/jOOQ/jOOQ/issues/9288

> Maybe this is just something that is better for the caller to deal with, i.e. `convertFrom(nullOnAnyNull(...))`.

Well, that version is just as good as the one you were suggesting, so there's no real value in adding those 5 additional methods in this case. In fact, even the conditional expression is acceptable in this case, given there's only one argument to the function, not possibly 22.
